### ISSUE TITLE
problem with sam.show_gene_expression

### ISSUE THREAD

**Issue by carlosj-rr** [atarashansky/self-assembling-manifold#33]
Hello. I ran into a problem when running sam.show_gene_expression() in order to plot the expression of a gene onto the clusters. I found a way around it, but it seemed the problem was in the function code. Please, let me know if you need any more information, I'm happy to help. Here's an extract of the code where I replicate the bug and find a workaround (plus some notes):

```
# sam-algorithm v 0.8.5 installed from pip:
$ pip show sam-algorithm
Name: sam-algorithm
Version: 0.8.5
Summary: The Self-Assembling-Manifold algorithm
Home-page: https://github.com/atarashansky/self-assembling-manifold
Author: Alexander J. Tarashansky
Author-email: tarashan@stanford.edu
[...]

>>> from samalg import SAM
>>> sam=SAM()
>>> sam.load_data("scrnaseq_data.csv.gz")
>>> sam.preprocess_data()
>>> sam.run()
RUNNING SAM
Iteration: 0, Convergence: 1.0
Iteration: 1, Convergence: 0.8693595227356725
Iteration: 2, Convergence: 0.018738846676010223
Computing the UMAP embedding...
Elapsed time: 162.4013922214508 seconds
>>> print(sam.adata.uns['ranked_genes'][:20])
['RHO' 'TULP1' 'SNHG11' 'MEG3' 'PCP2' 'RPGRIP1' 'SAG' 'CALM1' 'RCVRN'
 'PDE6G' 'GNAT1' 'A930011O12RIK' 'APOE' 'PDE6H' 'GNGT1' 'GLUL' 'PDC'
 'GNB1' 'TRPM1' 'PCP4']

>>> sam.scatter()
<matplotlib.axes._subplots.AxesSubplot object at 0x2b1f212f5450>
# things get funky in the next command suggested in the quickstart_guide.ipynb:
>>> sam.show_gene_expression(sam.adata.uns['ranked_genes'][0])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/mnt/storage/home/mi21309/miniconda3/envs/samenv/lib/python3.7/site-packages/sam_algorithm-0.8.5-py3.7.egg/samalg/__init__.py", line 732, in show_gene_expression
    a = self.adata.layers["X_knn_avg"][:, idx].toarray().flatten()
  File "/mnt/storage/home/mi21309/miniconda3/envs/samenv/lib/python3.7/site-packages/anndata/_core/aligned_mapping.py", line 148, in __getitem__
    return self._data[key]
KeyError: 'X_knn_avg'

# So the 'X_knn_avg' key is not working...in the sam.adata.layers["X_knn_avg"][:, idx].toarray().flatten() call.
# I try to understand a bit better the situation:

>>> type(sam.adata.layers)
<class 'anndata._core.aligned_mapping.Layers'> # Don't get this class..but them:
>>> sam.adata.layers
Layers with keys: X_disp
# it seems like the only key in that object is 'X_disp'.
# One way in which I succeeded to still get the colors was setting avg=False in the ".show_gene_expression" method.
# (I got that idea from the 'plotting_tutorial.ipynb' file)

>>> sam.show_gene_expression(sam.adata.uns['ranked_genes'][0],avg=False)
(<matplotlib.axes._subplots.AxesSubplot object at 0x2b33ca9892d0>, array([2.5849624, 3.4594316, 1.5849625, ..., 0.       , 5.044394 ,
       5.169925 ], dtype=float32))
>>> import matplotlib.pyplot as plt
# Then, plt.savefig() does the job!
```

**Comment by atarashansky** [comment]
Hi there, thanks for bringing this up. The tutorials are horribly out of date and I need to update them.

by default, the averaged expressions are no longer saved, so you’d need to save them by rerunning:

sam.dispersion_ranking_NN(save_avgs=True)

then, the plotting function should work.

**Comment by carlosj-rr** [comment]
Excellent, thanks! Otherwise excellent job on the algorithm!! works great!

**Comment by carlosj-rr** [comment]
Real quick, I tried the fix suggested and it crashes if the dataset is too big and there's not enough memory allocated. It's an obvious reason, and more likely my mistake than the developer's, but just wanted to let you know. With smaller datasets it works like a charm. Thanks!

Also, for anyone reading this, mind that the workaround that I presented (sam.show_gene_expression('gene', avg=False)) does NOT give the same results as the correction mentioned above by @atarashansky (which I assume is the correct output).
